### ISSUE TITLE
Question: Combinational vs Combinatorial

### ISSUE THREAD

**Issue by X-Illuminati** [amaranth-lang/amaranth#1301]
The [language guide](https://amaranth-lang.org/docs/amaranth/latest/guide.html) makes heavy reference to "combinatorial" logic.

In his reference text, _Digital Design: Principles and Practice_, Wakerly writes:

> A step _backward_ in MMI's introduction of PAL devices was their popularization of the word "combinatorial" to describe combinational circuits. _Combinational_ circuits have no memory -- their output at any time depends on the current input _combination_. For well-rounded computer engineers, the word "combinatorial" should conjure up vivid images of binomial coefficients, problem-solving complexity, and computer-science-great Donald Knuth.

[1, p. 343]

On the other hand, the Wikipedia article, [Combinational Logic](https://en.wikipedia.org/wiki/Combinational_logic), indicates that it is sometimes referred to as "combinatorial logic" and cites Maxfield's _FPGAs: World Class Designs_:

> Some folks prefer to say "combinational logic," while others favor "combinatorial logic."
> **1822:** England. Charles Babbage starts to build a mechanical calculating machine called the Difference Engine.

[[2, p. 16](https://books.google.com/books?id=kQuOKBSOz5QC&dq=%22combinatorial+logic%22&pg=PA16#v=onepage&q=%22combinatorial%20logic%22&f=false)]

So, perhaps the convention is not so clear cut.
Is Wakerly just being a curmudgeon? Is Maxfield actually citing something that Charles Babbage said? It isn't clear.

In any case, if you were in favor of changing the documentation to say "combinational" rather than "combinatorial", I would be happy to submit a pull-request for same.

---

[1] John F. Wakerly, _Digital Design: Principles and Practice_ (3rd Edition). Upper Saddle River, New Jersey, USA: Prentice hall, 2001.

[2] Clive Maxfield, _FPGAs: World Class Designs_. Newnes, 2009.

**Comment by cr1901** [comment]
I personally learned "combinational", but lots of people/resources (including a Xilinx app note I read literally 10 minutes ago) use "combinatorial". It's probably not worth the effort to be prescriptive about this.

**Comment by whitequark** [comment]
I legitimately don't know. It seems that the more senior ASIC engineers prefer "combinational" but there also doesn't seem to be a clear winner and people understand each other fine.

Ask a question on English.SE?

**Comment by whitequark** [comment]
From an [informal poll](https://mastodon.social/@whitequark/112251246846911479) it looks like nearly everyone is in either "I've never heard combinational" or "I've never heard combinatorial" camp, which is fascinating.

It looks like both VHDL and SystemVerilog standards use "combinational", but I'm not sure what lesson to take from that.

**Comment by tilk** [comment]
All the textbooks I use in teaching use "combinational". Also, Wikipedia has [an article on combinational logic](https://en.wikipedia.org/wiki/Combinational_logic), where they say it should not be confused with combinatory logic, which is a completely different thing.

**Comment by jfng** [comment]
I use "combinatorial", as it sounds closer to the word we use in french ("combinatoire"). I am happy with either one, as long as we use it consistently.

**Comment by whitequark** [comment]
It seems like:
- The use of "combinational" has no possible terminological objection, but some object to it on the aesthetic and familiarity basis.
- The use of "combinatorial" has terminological objection, as well as objection on the aesthetic and familiarity basis.

Clearly, going with "combinational" is preferred. @X-Illuminati Happy for you to do a PR and to add a note about this to our documentation style in docs/contrib!

**Comment by X-Illuminati** [comment]
Sure thing, I'm going to start working on the change and PR this afternoon. The recent edits to the wikipedia page do seem to support "combinational" from an editorial perspective and I see the talk page actually has some discussion on this as they previously renamed the article from Combinatorial Logic to Combinational Logic.

I enjoyed reading that mastodon thread, some funny reactions in there.
I have to profess a certain amount of Mandela Effect about this topic - even though I learned from the Wakerly text and specifically remembering his aside about the terminology, after 10 years or so I had convinced myself that "combinatorial" was the correct term. I heard a youtuber mention "combinational" a few years ago and was about to write a strongly worded comment, as one does, but then I looked it up again and found that I had simply misremembered it.
